sdks/java/fn-execution/src/main/java/org/apache/beam/sdk/fn/data/BeamFnDataOutboundAggregator.java

```diff
@@ -199,7 +210,14 @@ public void sendBufferedDataAndFinishOutboundStreams() {
           .setIsLast(true);
       entry.getValue().resetStats();
     }
+    if (collectElementsIfNoFlushes && !hasFlushedForBundle) {
+      return bufferedElements.build();
+    }
     outboundObserver.onNext(bufferedElements.build());
+    // This is now at the end of a bundle, so we reset hasFlushedForBundle to prepare for new
+    // bundles.
+    hasFlushedForBundle = false;
```

Comment:
the convertBufferForTransmission implicitly turns the flushThread to a no-op thread, so there should only be only one thread accessing hasFlushedForBundle.

Reply:
I'm not following since sendOrCollectBufferedDataAndFinishOutboundStreams() is invoked by the bundle processing thread and flush is invoked by the periodic flush thread.,x